src/components/LocationPermissionModal.tsx

```diff
@@ -0,0 +1,104 @@
+import React, {useEffect, useState} from 'react';
+import {Linking} from 'react-native';
+import {RESULTS} from 'react-native-permissions';
+import type {PermissionStatus} from 'react-native-permissions';
+import useLocalize from '@hooks/useLocalize';
+import useThemeStyles from '@hooks/useThemeStyles';
+import {getLocationPermission, requestLocationPermission} from '@pages/iou/request/step/IOURequestStepScan/LocationPermission';
+import ConfirmModal from './ConfirmModal';
+import * as Illustrations from './Icon/Illustrations';
+
+type LocationPermissionModalProps = {
+    /** A callback to call when the permission has been granted */
+    onGrant: () => void;
+
+    /** A callback to call when the permission has been denied */
+    onDeny: (permission: PermissionStatus) => void;
+
+    /** Should start the permission flow? */
+    startPermissionFlow: boolean;
+
+    /** Reset the permission flow */
+    resetPermissionFlow: () => void;
+};
+
+function LocationPermissionModal({startPermissionFlow, resetPermissionFlow, onDeny, onGrant}: LocationPermissionModalProps) {
+    const [hasError, setHasError] = useState(false);
+    const [showModal, setShowModal] = useState(false);
+
+    const styles = useThemeStyles();
+    const {translate} = useLocalize();
+
+    useEffect(() => {
+        if (!startPermissionFlow) {
+            return;
+        }
+
+        getLocationPermission().then((status) => {
+            if (status === RESULTS.GRANTED || status === RESULTS.LIMITED) {
+                return onGrant();
+            }
+
+            setShowModal(true);
+            setHasError(status === RESULTS.BLOCKED);
+        });
+        // eslint-disable-next-line react-hooks/exhaustive-deps -- We only want to run this effect when startPermissionFlow changes
+    }, [startPermissionFlow]);
+
+    const errorHandler = (cb: () => void) => () => {
+        if (hasError) {
+            Linking.openSettings?.();
+            setShowModal(false);
+            setHasError(false);
+            resetPermissionFlow();
+            return;
+        }
+        cb();
+    };
+
+    const onConfirm = errorHandler(() => {
```

Comment:
I am going to leave a comment here but this applies everywhere. These don't seem to be following the convention,  these should be named for what it does, not what it handles: https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#event-handlers